### PR TITLE
Added Kubernetes Logging Plugin for OMS Agent

### PR DESCRIPTION
**Pull request by JasonStein**
Plugin for sending k8s logs to OMS agent based on namespace.

### CODE REVIEW

File: source/code/plugin/in_kube_logs.rb [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                            end
+                        end    
+                    rescue  => errorStr
+                        $log.warn line.dump, error: errorStr.to_s
+                        $log.debug_backtrace(e.backtrace)
+                    end
+                end
+                writeLogQueryState(newLogQueryState)
+            else
+                record = {}
```
Comment (OriYosefiMSFT) [inline source/code/plugin/in_kube_logs.rb]: when is this called? Why call the router with empty record?

**Comment by OriYosefiMSFT** [comment]
@samisms , @keikhara - FYI

File: source/code/plugin/KubernetesApiClient.rb [hunk 2]
```diff
@@ -1,128 +1,173 @@
...
+            def isNodeMaster
+                return @@IsNodeMaster if !@@IsNodeMaster.nil?
+                @@IsNodeMaster = false
+                begin
+                    allNodesInfo = JSON.parse(getKubeResourceInfo('nodes').body)
+                    if !allNodesInfo.nil? && !allNodesInfo.empty?
+                        thisNodeName = OMS::Common.get_hostname
+                        allNodesInfo['items'].each do |item|
+                            if item['metadata']['name'].casecmp(thisNodeName) == 0
+                                if item['metadata']['labels']["kubernetes.io/role"].to_s.include?("master") && item['metadata']['labels']["kubernetes.io/hostname"].to_s.split('-').last == '0'
```
Comment (samisms) [inline source/code/plugin/KubernetesApiClient.rb]: Can you add a check for the Windows / Linux environment variable. Like we discussed the last time pinning it to the master doesnt work for pure linux environment, in which case it needs to be one of the agent nodes. For the time being we are pinning to node 0 .

File: installer/conf/container.conf [hunk 3]
```diff
@@ -94,6 +94,13 @@
 	run_interval 60s
 </source>
 
+#Kubernetes logs
+<source>
+	type kubelogs
+	tag oms.api.KubeLogs
+	run_interval 5s
```
Comment (samisms) [inline installer/conf/container.conf]: 5s run interval might be too aggressive for scale. Can you do a scale test of 1000/1000 of your poc pods logging every second ?

File: source/code/plugin/in_kube_logs.rb [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                                    dateTime = DateTime.parse(lines[i].split(" ").first)               
+                                    if (dateTime.to_time - timeStamp.to_time) > 0.0
+                                        index = i
+                                        break
+                                    end
+                                end
+            
+                                if index >= 0
+                                    for i in index...lines.count
+                                        record['Namespace'] = pod['metadata']['namespace']
```
Comment (samisms) [inline source/code/plugin/in_kube_logs.rb]: Can you also add cluster name and computer to this record ?

**Comment by samisms** [comment]
Code merged through https://github.com/Microsoft/Docker-Provider/pull/46#event-1418753838

**Pull request closed without merging**